motion_primitives_forward_controller/src/motion_primitives_forward_controller.cpp

```diff
@@ -0,0 +1,554 @@
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  reset_command_interfaces();
+  RCLCPP_INFO(get_node()->get_logger(), "Controller activated");
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn MotionPrimitivesForwardController::on_deactivate(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  reset_command_interfaces();
+  RCLCPP_INFO(get_node()->get_logger(), "Controller deactivated");
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::return_type MotionPrimitivesForwardController::update(
+  const rclcpp::Time & time, const rclcpp::Duration & /*period*/)
+{
+  // read the status from the state interface
+  auto opt_value_execution = state_interfaces_[0].get_optional();
+  if (!opt_value_execution.has_value())
+  {
+    RCLCPP_ERROR(
+      get_node()->get_logger(), "State interface 0 (execution_state) returned no value.");
+    return controller_interface::return_type::ERROR;
+  }
+  execution_status_ = static_cast<int8_t>(std::round(opt_value_execution.value()));
+
+  switch (execution_status_)
+  {
+    case ExecutionState::IDLE:
+      // RCLCPP_INFO(get_node()->get_logger(), "Execution state: IDLE");
+      print_error_once_ = true;
+      break;
+    case ExecutionState::EXECUTING:
+      // RCLCPP_INFO(get_node()->get_logger(), "Execution state: EXECUTING");
+      if (!was_executing_)
+      {
+        was_executing_ = true;
+      }
+      print_error_once_ = true;
+      break;
+
+    case ExecutionState::SUCCESS:
+      // RCLCPP_INFO(get_node()->get_logger(), "Execution state: SUCCESS");
+      print_error_once_ = true;
+      break;
+
+    case ExecutionState::STOPPED:
+      // RCLCPP_INFO(get_node()->get_logger(), "Execution state: STOPPED");
+      print_error_once_ = true;
+      break;
+
+    case ExecutionState::ERROR:
+      if (print_error_once_)
+      {
+        RCLCPP_ERROR(get_node()->get_logger(), "Execution state: ERROR");
+        print_error_once_ = false;
+      }
+      break;
+
+    default:
+      RCLCPP_ERROR(
+        get_node()->get_logger(), "Error: Unknown execution status: %d", execution_status_);
+      return controller_interface::return_type::ERROR;
+  }
+
+  auto opt_value_ready = state_interfaces_[1].get_optional();
+  if (!opt_value_ready.has_value())
+  {
+    RCLCPP_ERROR(
+      get_node()->get_logger(), "State interface 1 (ready_for_new_primitive) returned no value.");
+    return controller_interface::return_type::ERROR;
+  }
+  uint8_t ready_for_new_primitive = static_cast<int8_t>(std::round(opt_value_ready.value()));
+
+  if (!moprim_queue_.empty())  // check if new command is available
+  {
+    switch (ready_for_new_primitive)
+    {
+      case ReadyForNewPrimitive::NOT_READY:
+      {
+        return controller_interface::return_type::OK;
+      }
+      case ReadyForNewPrimitive::READY:
+      {
+        if (!set_command_interfaces())
```

Comment:
This is where you potentially block the control thread as this will wait for aquiring the lock for accessing the queue.